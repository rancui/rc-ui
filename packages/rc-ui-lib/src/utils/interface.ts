import type { CSSProperties } from 'react';

export interface BaseTypeProps {
  style?: CSSProperties;
  className?: string;
  children?: React.ReactNode;
}

/** 指定挂载的节点 */
export type TeleportType = HTMLElement | (() => HTMLElement) | null;
/** 带有displayName的组件 */
export type WithDisplayNameReactElement = React.ReactElement & { type: { displayName: string } };
