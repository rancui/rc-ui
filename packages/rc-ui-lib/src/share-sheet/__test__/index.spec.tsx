import React from 'react';
import { cleanup, fireEvent, render } from '@testing-library/react';
import { sleep } from '../../../tests/utils';
import { ShareSheet } from '..';

const optionsMuli = [
  [
    { name: '微信', icon: 'wechat' },
    { name: '朋友圈', icon: 'wechat-moments' },
    { name: '微博', icon: 'https://img.yzcdn.cn/vant/share-sheet-weibo.png' },
    { name: 'QQ', icon: 'qq' },
  ],
  [
    { name: '复制链接', icon: 'link' },
    { name: '分享海报', icon: 'poster' },
    { name: '二维码', icon: 'qrcode' },
    { name: '小程序码', icon: 'weapp-qrcode' },
  ],
];

describe('ShareSheet', () => {
  afterEach(() => {
    cleanup();
    jest.restoreAllMocks();
  });

  test('should render cancel text when using cancel-text prop', async () => {
    const { rerender } = render(<ShareSheet visible cancelText="foo" />);

    expect(document.querySelector('.rc-share-sheet__cancel')).toMatchSnapshot();

    rerender(<ShareSheet visible cancelText="" />);
    expect(document.querySelector('.rc-share-sheet__cancel')).toBeFalsy();
  });

  test('should render description and title when using description prop and title prop', async () => {
    const { rerender } = render(<ShareSheet visible description="foo" title="hoo" />);

    expect(document.querySelector('.rc-share-sheet__description')).toMatchSnapshot();
    expect(document.querySelector('.rc-share-sheet__title')).toMatchSnapshot();
    rerender(<ShareSheet visible description="" title="hoo" />);
    expect(document.querySelector('.rc-share-sheet__description')).toBeFalsy();
  });

  test('should allow to custom the className of option', () => {
    render(
      <ShareSheet
        visible
        options={[{ name: 'Link', icon: 'link', className: 'foo', description: 'hoo' }]}
      />,
    );

    const option = document.querySelector('.rc-share-sheet__option');

    expect(option.classList.contains('foo')).toBeTruthy();
  });
  // description

  test('should render multi Line when option item is array', () => {
    render(<ShareSheet visible options={optionsMuli} />);

    expect(document.querySelector('.rc-share-sheet')).toMatchSnapshot();
  });

  test('should emit select event when an option is clicked', async () => {
    const onSelect = jest.fn();
    render(
      <ShareSheet visible options={[{ icon: 'wechat', name: 'wechat' }]} onSelect={onSelect} />,
    );

    await fireEvent.click(document.querySelector('.rc-share-sheet__option'));
    expect(onSelect).toHaveBeenCalledWith({ icon: 'wechat', name: 'wechat' }, 0);
  });

  test('should emit cancel event when the cancel button is clicked', async () => {
    const onCancel = jest.fn();
    render(<ShareSheet visible onCancel={onCancel} />);

    await fireEvent.click(document.querySelector('.rc-share-sheet__cancel'));

    expect(onCancel).toHaveBeenCalledTimes(1);
  });

  test('should emit click-overlay event when overlay is clicked', async () => {
    const onCancel = jest.fn();
    render(<ShareSheet visible cancelText="foo" onCancel={onCancel} />);

    await sleep();

    const overlay = document.querySelector('.rc-overlay');
    await fireEvent.click(overlay);
    expect(onCancel).toHaveBeenCalledTimes(1);
  });

  test('should have "rc-popup--round" class when setting the round prop', async () => {
    render(<ShareSheet visible round />);

    expect(document.querySelector('.rc-popup--round')).toBeTruthy();
  });
});
