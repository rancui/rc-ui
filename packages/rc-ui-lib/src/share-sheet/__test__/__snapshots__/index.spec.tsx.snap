// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ShareSheet should render cancel text when using cancel-text prop 1`] = `
<button
  class="rc-share-sheet__cancel"
  type="button"
>
  foo
</button>
`;

exports[`ShareSheet should render description and title when using description prop and title prop 1`] = `
<div
  class="rc-share-sheet__description"
>
  foo
</div>
`;

exports[`ShareSheet should render description and title when using description prop and title prop 2`] = `
<div
  class="rc-share-sheet__title"
>
  hoo
</div>
`;

exports[`ShareSheet should render multi Line when option item is array 1`] = `
<div
  class="rc-popup rc-popup--round rc-popup--bottom rc-safe-area-bottom rc-share-sheet"
  style="z-index: 2000; transition-duration: 300ms;"
>
  <div
    class="rc-share-sheet__options"
  >
    <div
      class="rc-share-sheet__option"
      role="button"
      tabindex="0"
    >
      <img
        alt="share sheet icon"
        class="rc-share-sheet__icon"
        src="https://img.yzcdn.cn/vant/share-sheet-wechat.png"
      />
      <span
        class="rc-share-sheet__name"
      >
        微信
      </span>
    </div>
    <div
      class="rc-share-sheet__option"
      role="button"
      tabindex="0"
    >
      <img
        alt="share sheet icon"
        class="rc-share-sheet__icon"
        src="https://img.yzcdn.cn/vant/share-sheet-wechat-moments.png"
      />
      <span
        class="rc-share-sheet__name"
      >
        朋友圈
      </span>
    </div>
    <div
      class="rc-share-sheet__option"
      role="button"
      tabindex="0"
    >
      <img
        alt="share sheet icon"
        class="rc-share-sheet__icon"
        src="https://img.yzcdn.cn/vant/share-sheet-weibo.png"
      />
      <span
        class="rc-share-sheet__name"
      >
        微博
      </span>
    </div>
    <div
      class="rc-share-sheet__option"
      role="button"
      tabindex="0"
    >
      <img
        alt="share sheet icon"
        class="rc-share-sheet__icon"
        src="https://img.yzcdn.cn/vant/share-sheet-qq.png"
      />
      <span
        class="rc-share-sheet__name"
      >
        QQ
      </span>
    </div>
  </div>
  <div
    class="rc-share-sheet__options rc-share-sheet__options--border"
  >
    <div
      class="rc-share-sheet__option"
      role="button"
      tabindex="0"
    >
      <img
        alt="share sheet icon"
        class="rc-share-sheet__icon"
        src="https://img.yzcdn.cn/vant/share-sheet-link.png"
      />
      <span
        class="rc-share-sheet__name"
      >
        复制链接
      </span>
    </div>
    <div
      class="rc-share-sheet__option"
      role="button"
      tabindex="0"
    >
      <img
        alt="share sheet icon"
        class="rc-share-sheet__icon"
        src="https://img.yzcdn.cn/vant/share-sheet-poster.png"
      />
      <span
        class="rc-share-sheet__name"
      >
        分享海报
      </span>
    </div>
    <div
      class="rc-share-sheet__option"
      role="button"
      tabindex="0"
    >
      <img
        alt="share sheet icon"
        class="rc-share-sheet__icon"
        src="https://img.yzcdn.cn/vant/share-sheet-qrcode.png"
      />
      <span
        class="rc-share-sheet__name"
      >
        二维码
      </span>
    </div>
    <div
      class="rc-share-sheet__option"
      role="button"
      tabindex="0"
    >
      <img
        alt="share sheet icon"
        class="rc-share-sheet__icon"
        src="https://img.yzcdn.cn/vant/share-sheet-weapp-qrcode.png"
      />
      <span
        class="rc-share-sheet__name"
      >
        小程序码
      </span>
    </div>
  </div>
  <button
    class="rc-share-sheet__cancel"
    type="button"
  >
    取消
  </button>
</div>
`;
